gems/smithy-schema/lib/smithy-schema/document/serializer.rb

```diff
@@ -0,0 +1,283 @@
+              shape = @type_registry[data['__type']]
+              Data.new(format_document_data(shape, data, discriminator: true), discriminator: shape.id)
+            else
+              Data.new(serialize_untyped(data))
+            end
+          end
+        end
+
+        # Serializes a document data with optional formatting.
+        # @param [Data] document The document to serialize
+        # @param [Hash] opts serialization options
+        # @option opts [Boolean] :use_timestamp_format Whether to use the
+        #  `timestampFormat` trait or ignore it. The `timestampFormat` trait
+        #   is ignored by default.
+        # @option opts [Boolean] :use_json_name Whether to use `jsonName` trait
+        #   or just member name. The `jsonName` trait is ignored by default.
+        # @return [Hash] Serialized document data
+        def serialize_document(document, opts = {})
+          validate_document(document)
+
+          opts[:discriminator] = true
+          format_document_data(resolve_shape(document), document.data, opts)
+        end
+
+        private
+
+        def serialize_untyped(values)
+          return if values.nil?
+
+          case values
+          when Time
+            values.to_i # timestamp format is "epoch-seconds" by default
+          when Hash
+            values.each_with_object({}) do |(k, v), h|
+              h[k.to_s] = serialize_untyped(v)
+            end
+          when Array
+            values.map { |d| serialize_untyped(d) }
+          else
+            values
+          end
+        end
+
+        def shape(ref, values, opts = {}) # rubocop:disable Metrics/CyclomaticComplexity
+          return if values.nil?
+
+          case ref.shape
+          when BlobShape      then blob(values, opts)
+          when DocumentShape  then document(values, opts)
+          when FloatShape     then float(values, opts)
+          when ListShape      then list(ref, values, opts)
+          when MapShape       then map(ref, values, opts)
+          when StructureShape then structure(ref, values, opts)
+          when TimestampShape then timestamp(ref, values, opts)
+          when UnionShape     then union(ref, values, opts)
+          else values
+          end
+        end
+
+        def blob(value, opts)
+          return value if opts[:discriminator] # blob is already encoded
+
+          Base64.strict_encode64(value.is_a?(String) ? value : value.read)
+        end
+
+        def document(values, opts)
+          return values unless typed_document?(values)
+
+          shape =
+            if values.is_a?(Smithy::Schema::Structure)
+              @type_registry.shape_by_type(values.class)
+            else
+              @type_registry[values['__type']]
+            end
+          format_document_data(shape, values, opts)
+        end
+
+        def typed_document?(values)
+          (values.is_a?(Smithy::Schema::Structure) && @type_registry.shape_by_type(values.class)) ||
```

Comment:
Isn't this already checked? And wouldn't this always be true if it was a structure, because it would already be registered?